### PR TITLE
[system-probe] Start of dynamic instrumentation system probe module

### PR DESCRIPTION
**Pull request by grantseltzer**
### What does this PR do?

This introduces the scaffolding for the new dynamic instrumentation system-probe module. This will be a module used by the dynamic-instrumentation product for system level tracing of user applications. This belongs in the system-probe as it will soon need higher level permissions (CAP_SYS_ADMIN, CAP_BPF) for loading and attaching bpf programs, and inspecting inside container filesystems.

### Motivation

This PR simply creates the module in code, and gives it the scaffolding for reading a static configuration file. The motivation for this is to break up the large amount of code that will need to be placed here into smaller PRs.

### Additional Notes

- This also is specified via go build tags to only build on Linux which is required for the bpf features it will consume.

### Describe how to test/QA your changes

- Build the system probe: `inv system-probe.build`
- Run the system-probe `sudo bin/system-probe/system-probe`

You should see it's disabled:
```
2023-03-08 13:37:33 PST | SYS-PROBE | INFO | (cmd/system-probe/api/module/loader.go:50 in Register) | module dynamic_instrumentation disabled
2023-03-08 13:37:33 PST | SYS-PROBE | INFO | (cmd/system-probe/app/run.go:123 in run) | system probe successfully started
```

You can run it enabled by setting `DD_DYNAMIC_INSTRUMENTATION_ENABLED=1` or adding the following section to your system-probe config:

```
####################################
## Dynamic Instrumentation Config ##
###################################

dynamic_instrumentation:
  ## @param enabled - boolean - optional - default: false
  ## @env DD_DYNAMIC_INSTRUMENTATION_ENABLED - boolean -optional - default: false
  ## Set to true to enable the User Tracer Module of the System Probe
  #
  enabled: true
```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by guyarb — COMMENTED** [review]
You should modify the CODEOWNERS files to assign your new files to be owned by your team for future reviews

**Comment by grantseltzer** [comment]
@guyarb Good idea!

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -247,6 +247,7 @@
 /pkg/config/remote/service/meta/        @DataDog/remote-config @DataDog/software-integrity-and-trust
 /pkg/diagnose/                          @Datadog/container-integrations
 /pkg/diagnose/connectivity/             @DataDog/agent-shared-components
+/pkg/dynamic-instrumentation/           @DataDog/debugger
```
Comment (sgnn7) [inline .github/CODEOWNERS:250]: Generally I think for Go, multi-term package names should be all lower-case with no special characters so I think this should be `/pkg/dynamicinstrumentation/` or a composite if it makes sense (`pkg/instrumentation/dynamic`). With that said, our codebase isn't quite consistent about this but we should strive to make it better moving forward.

File: pkg/config/system_probe.go [hunk 2]
```diff
@@ -131,6 +133,9 @@ func InitSystemProbeConfig(cfg Config) {
 	cfg.BindEnvAndSetDefault(join(spNS, "zypper_repos_dir"), suffixHostEtc(defaultZypperReposDirSuffix), "DD_ZYPPER_REPOS_DIR")
 	cfg.BindEnvAndSetDefault(join(spNS, "attach_kprobes_with_kprobe_events_abi"), false, "DD_ATTACH_KPROBES_WITH_KPROBE_EVENTS_ABI")
 
+	// User Tracer
+	cfg.BindEnvAndSetDefault(join(diNS, "enabled"), true, "DD_DYNAMIC_INSTRUMENTATION_ENABLED")
```
Comment (guyarb) [inline pkg/config/system_probe.go:137]: does it make sense to enable the (new) feature by default?

Comment (grantseltzer) [reply pkg/config/system_probe.go:137]: Nope - i'll disable

File: .github/CODEOWNERS (hunk 1 above)
Comment (grantseltzer) [reply .github/CODEOWNERS]: That works for me, fixed.

**Comment by grantseltzer** [comment]
> I think this PR will not trigger the gitlab CI because it originates from a fork and not from a local branch of the repo (cc @KSerrania)

Ah, my mistake. I can open a fresh PR if need be.

**Comment by KSerrania** [comment]
> I think this PR will not trigger the gitlab CI because it originates from a fork and not from a local branch of the repo (cc @KSerrania)

Yes, that is indeed the case. I could probably make it run once manually, but if more changes are needed this will be slow, so opening a new PR with a branch in the repo will likely be faster.

**Comment by grantseltzer** [comment]
@KSerrania @paulcacheux  Ok I've opened https://github.com/DataDog/datadog-agent/pull/16034 to replace this one. Closing here.

**Pull request closed without merging**